### ISSUE TITLE
AutoLoad Singleton using Nim?

### ISSUE THREAD

**Issue by zetashift** [pragmagic/godot-nim#38]
I currently have 2 nim, player.nim and game_state.nim files containing a KinematicBody2D(as Player classname) and a Node2D(as a GameState classname).
I also have a Global script as an AutoLoad script, however in my corresponding 'global.nim' which tracks the current gameState and player, I can't set/get anything without introducing recursive importing problems, which is obvious since Global imports player.nim and game_state.nim but those 2 files also import global so they can set it in their `ready()` methods.

What would be the Nim way(TM) without introducing importing problems?

**Comment by zetashift** [comment]
Might be related: https://github.com/godotengine/godot/issues/20880

**Comment by endragor** [comment]
GDNative AutoLoads do work fine in 3.1 for me.

As about breaking the dependency cycle, I would need more details to suggest the best way to resolve that. You can always do it like in any OOP system - you can break the cycle with either inheritance or aggregation. Make a GlobalInterface or GlobalBase class providing the methods that are needed by Player and GameState, inherit Global from it and override the methods. This way, everything depends on GlobalBase and GlobalBase depends on nothing, no cycles. If GlobalBase methods need to have Player and GameState in their signatures, then make base classes for those, too.

There could be ways to break the dependency without introducing more classes, but that depends on the actual use case. If sole purpose of Global is to provide GameState and Player instances, why it doesn't create them? It could look like this:

```nim
# global.nim
import godot, node
import player, game_state

gdobj Global of Node:
  var player* {.gdExport.}: Player
  var gameState* {.gdExport.}: GameState

  method ready() =
    player = gdnew[Player]()
    player.name = "Player"
    addChild(player)

    gameState = gdnew[GameState]()
    gameState.name = "GameState"
    addChild(gameState)
```

```nim
# game_state.nim
gdobj GameState of Node2D:
  # ...

proc getGameState*(n: Node): GameState =
  n.getNode("/root/Global/GameState") as GameState
```

```nim
# game_state.nim
gdobj Player of KinematicBody2D:
  # ...

proc getPlayer*(n: Node): Player =
  n.getNode("/root/Global/Player") as Player
```

**Comment by zetashift** [comment]
@endragor thanks! Your example clarified a lot. Basically my Global singleton will provide the instances indeed so I don't have to deal with getting nodes based on structure. 

I currently with your suggestion have this
```nim
#snippet of global.nim
gdobj Global of Node:
    var player* {.gdExport.}: Player
    var gameState* {.gdExport.}: GameState

    method ready*() =
        player = gdnew[Player]()
        player.name= "Player"
        addChild(player)
        
        gameState = gdnew[GameState]()
        gameState.name = "GameState"
        addChild(gameState)
```
and
```nim
#game_state.nim
gdobj Global of Node:
    var player* {.gdExport.}: Player
    var gameState* {.gdExport.}: GameState

    method ready*() =
        player = gdnew[Player]()
        player.name= "Player"
        addChild(player)
        
        gameState = gdnew[GameState]()
        gameState.name = "GameState"
        addChild(gameState)
```

And in player nim I calculate if the player is out of bounds and call the `endGame` proc called from the Globals GameState instance

**Comment by zetashift** [comment]
The `Script does not inherit a Node: res://Scripts/Global.gdns` got solved by manually editing the Global.gdns to include the nim.gdnlib. For some reason the editor didn't add it

**Comment by zetashift** [comment]
Alright after some messing around, I ended up where I started just using nodepaths and having an empty `global.nim`
the gdnew approach gave me an error that it couldn't find an animation node, which is a child of the Player class. 
I'm following the godot course that's why I am kind of stuck on this approach. It also doesn't help that I don't know much about advanced OOP concepts. 

What I'm trying to do is nimifying this: https://github.com/DiscoveringGodot/3_Hoppy_Days/tree/master/Scripts
The `Global.gd` holds variables which are set in their corresponding classes, in their `ready()` methods. I saw a C# example which did the same, 1:1 translation of it results in the recursive dependency problem.

**Comment by endragor** [comment]
I wouldn't do it the way proposed there even if it wasn't Nim. Such things may work for a small demo/example where you can keep everything in your head, but they will turn your codebase into a mess if it gets larger.

This is roughly the same sample as I gave earlier, just with `load()` instead of `gdnew` and the dependencies are supplied to the creation procedure, so the classes don't have to depend on Global (called `Game` here). 

```nim
gdobj Game of Node:
  var player* {.gdExport.}: Player
  var gameState* {.gdExport.}: GameState

  method ready() =
    gameState = newGameState() # change GameState not to depend on Player
    gameState.name = "GameState"
    addChild(gameState)

    player = newPlayer(gameState)
    player.name = "Player"
    addChild(player)
```

```nim
# player.nim
import godot, node, kinematic_body_2d
import game_state

gdobj Player of KinematicBody2D:
  var gameState: GameState
  # ...

proc newPlayer(gameState: GameState): Player =
  result = load("res://.../Player.tscn") as Player
  result.gameState = gameState

proc getPlayer*(n: Node): Player =
  n.getNode("/root/Game/Player") as Player
```

**Comment by zetashift** [comment]
Yeah I was wondering why all the other tutorials I looked at looking for singleton discussion didn't mention any of this.

Currently have this as my `newGameState()`
```nim
#game_state.nim
proc newGameState*(): GameState =
    result = load("res://Scenes/Level1.tscn") as GameState

proc getGameState*(n: Node): GameState =
    n.getNode("/root/Global/GameState") as GameState
```
And getting the following error
`0:00:00:0471 - Failed to cast object to GameState
Traceback (most recent call last)
godotmacros.nim(573)     methFunc
global.nim(10)           ready
game_state.nim(14)       newGameState
godotnim.nim(337)        as

----------
Type:Error
Description: 
Time: 0:00:00:0471
C Error: Failed to cast object to GameState
Traceback (most recent call last)
godotmacros.nim(573)     methFunc
global.nim(10)           ready
game_state.nim(14)       newGameState
godotnim.nim(337)        as

C Source: /home/rishi/.nimble/pkgs/godot-0.7.17/nim/godotnim.nim:336
C Function: 
`
the GameState inherits from Node2D and the Level1.tscn is a scene file containing a root Node2D with the tilemap/player/camera as children.

**Comment by endragor** [comment]
The root of Level1.tscn has to have GameState.gdns script attached to it.

**Comment by zetashift** [comment]
It does! The root node of Level1.tscn is called Level1(and is a Node2D) and attached to it a `GameState.gdns`
